src/template/wasm2c.declarations.c

```diff
@@ -89,20 +89,22 @@ static inline bool func_types_eq(const wasm_rt_func_type_t a,
   (CHECK_CALL_INDIRECT(table, ft, x),       \
    DO_CALL_INDIRECT(table, t, x, __VA_ARGS__))
 
-#ifdef SUPPORT_MEMORY64
+#if __has_builtin(__builtin_add_overflow)
+#define add_overflow(a, b, resptr) __builtin_add_overflow(a, b, resptr)
+#else /* MSVC */
+static inline bool add_overflow(uint64_t a, uint64_t b, uint64_t* resptr) {
+  return _addcarry_u64(0, a, b, resptr);
+}
+#endif
+
 #define RANGE_CHECK(mem, offset, len)              \
```

Comment:
Unfortunately defaulting to a 64-bit check will slow down the performance of bounds checked wasm2c for 32-bit memories. Firefox uses the bounds checked wasm2c for 32-bit devices, and so it is perf sensitive to this. I don't know if this is addressed in a future PR, but this particular PR would be a perf problem from the Firefox use case.

Also, I'll note that defaulting to the 64-bit RANGE check is fine for memcpy, tables, etc. as I don't think it will affect performance, the only concern is the 64-bit RANGE on accesses to 32-bit linear memories

Reply:
Thanks for these thoughtful (and well-taken) comments. I believe #2507 will nail this for you (by preserving the current RANGE_CHECK on 32-bit, default-page-size memories), so, how about we wait to get alignment on both #2506 and #2507 and then land them at the same time.

I should say that even the current RANGE_CHECK uses 64-bit arithmetic:

```
#define RANGE_CHECK(mem, offset, len)               \
  if (UNLIKELY(offset + (uint64_t)len > mem->size)) \
    TRAP(OOB);
```

... but the difference is that RANGE_CHECK64 does an explicit check for 64-bit overflow. I wish I had the benchmarking infrastructure to promise you it won't affect performance on 32-bit x86 but... safer to wait for #2507 which lets you keep the same code.